### ISSUE TITLE
Allow top-level functions to be annotated

### ISSUE THREAD

**Issue by JanPokorny** [InsertKoinIO/koin-annotations#68]
**Is your feature request related to a problem? Please describe.**

I have many files that define `HttpClient`s like so:

```kotlin
// FooHttpClient.kt
fun buildFooHttpClient(config: FooConfig) = HttpClient(CIO) { /* ... */ }

// BarHttpClient.kt
fun buildBarHttpClient(config: BarConfig) = HttpClient(CIO) { /* ... */ }

// ...etc
```

...and a module like so:

```kotlin
val httpClientsModule = module {
	singleOf(::buildFooHttpClient) { named("Foo") }
	singleOf(::buildBarHttpClient) { named("Bar") }
	// ...etc
}
```

I am porting the codebase to use Koin annotations.

**Describe the solution you'd like**

Analogous to how class definitions work, it seemed logical that I would just annotate the builder functions with `@Single(binds = [HttpClient::class]) @Named("Foo")`, and then `@Module @ComponentScan(...) class HttpClientsModule` would pick them up.

This is not the case. Annotating free standing functions generates no code.

**Describe alternatives you've considered**

I could put all the builder functions as methods to a single class and annotate it as `@Module`, which works but requires me to keep all the definitions in a single file. There is no way to split that class without involving a lot of boilerplate code, the very kind I try to avoid by using koin-annotations.

**Target Koin project**
koin-annotations

**Comment by JanPokorny** [comment]
@arnaudgiuliani I'm not sure what you mean. Could you provide an example?

**Comment by arnaudgiuliani** [comment]
I mean, you can declare functions from a class:

```kotlin

@Module
class HttpModule {

    @Single
    @Qualifier("foo")
    fun buildFooHttpClient(config: FooConfig) = HttpClient(CIO) { /* ... */ }

    @Single
    @Qualifier("bar")
    fun buildBarHttpClient(config: BarConfig) = HttpClient(CIO) { /* ... */ }
}
```

**Comment by JanPokorny** [comment]
@arnaudgiuliani Yes, I'm aware, I mention this in the issue description. This approach won't let me keep the project structure, I would need to move all the functions to a single file. I could also create a module per each function, but then I'd still need to list all the modules somewhere, defeating the point.

I'm convinced no elegant approach is possible at the moment and that implementing the proposal would be helpful.

**Comment by arnaudgiuliani** [comment]
Ok interesting. If I wrap up your case here, is to tag a function as a Koin component directly. No module class needed.

**Comment by JanPokorny** [comment]
@arnaudgiuliani Yes. It would behave analogically to annotated classes (ending up in a @ComponentScan module or the defaultModule).

**Comment by arnaudgiuliani** [comment]
will be in next milestone: 1.2.0 👍
